gcloud/monitoring/query.py

```diff
@@ -0,0 +1,661 @@
+        new_query._alignment_period_seconds = seconds + 60 * (minutes +
+                                                              60 * hours)
+        return new_query
+
+    def reduce(self, cross_series_reducer, *group_by_fields):
+        """Copy the query and add cross-series reduction.
+
+        Cross-series reduction combines time series by aggregating their
+        data points.
+
+        For example, you could request an aggregated time series for each
+        combination of project and zone as follows::
+
+            query = query.reduce(Reducer.REDUCE_MEAN,
+                                 'resource.project_id', 'resource.zone')
+
+        :type cross_series_reducer: string
+        :param cross_series_reducer:
+            The approach to be used to combine time series. For example:
+            ``Reducer.REDUCE_MEAN``. See :class:`Reducer` and the descriptions
+            of the `supported reducers`_.
+
+        :type group_by_fields: strings
+        :param group_by_fields:
+            Fields to be preserved by the reduction. For example, specifying
+            just ``"resource.zone"`` will result in one time series per zone.
+            The default is to aggregate all of the time series into just one.
+
+        :rtype: :class:`Query`
+        :returns: The new query object.
+
+        .. _supported reducers:
+            https://cloud.google.com/monitoring/api/ref_v3/rest/v3/\
+            projects.timeSeries/list#Reducer
+        """
+        new_query = self.copy()
+        new_query._cross_series_reducer = cross_series_reducer
+        new_query._group_by_fields = group_by_fields
+        return new_query
+
+    def iter(self, headers_only=False, _page_size=None):
+        """Yield all time series objects selected by the query.
+
+        Note that the :class:`Query` object itself is an iterable, such that
+        the following are equivalent::
+
+            for timeseries in query: ...
+            for timeseries in query.iter(): ...
+
+        :type headers_only: boolean
+        :param headers_only:
+             Whether to omit the point data from the time series objects.
+
+        :type _page_size: integer or None
+        :param _page_size:
+            An optional positive number specifying the maximum number of
+            points to return per page. This can be used to control how far
+            the iterator reads ahead.
+
+        :rtype: iterator over :class:`~gcloud.monitoring.timeseries.TimeSeries`
+        :returns: Time series objects, containing points ordered from oldest
+            to newest.
+        :raises: :exc:`ValueError` if the query time interval has not been
+            specified.
+        """
+        if self._end_time is None:
+            raise ValueError('Query time interval not specified.')
+
+        path = '/projects/{project}/timeSeries/'.format(
+            project=self._client.project)
+
+        def _fragments():
```

Comment:
We have gone out of our way to avoid conditionally defined classes / functions in this manner.

This could be re-factored into an instance method

```python
    def _iter_fragments(self, path, page_size, headers_only):
        ...
```

and accomplish the same task without the need for an unneeded closure.